File: 프로그래머스/1/131128. 숫자 짝꿍/숫자 짝꿍.kt
class Solution {
    fun solution(X: String, Y: String): String {
        val xCount = IntArray(10)
        val yCount = IntArray(10)

        // X 문자열에서 각 숫자 개수 세기
        for (ch in X) {
            xCount[ch - '0']++
        }

        // Y 문자열에서 각 숫자 개수 세기
        for (ch in Y) {
            yCount[ch - '0']++
        }

        val sb = StringBuilder()

        // 큰 숫자부터 공통으로 있는 만큼 추가
        for (i in 9 downTo 0) {
            val count = minOf(xCount[i], yCount[i])
            repeat(count) {
                sb.append(i)
            }
        }

        // 결과 처리
        val result = sb.toString()
        return when {
            result.isEmpty() -> "-1"
            result[0] == '0' -> "0"  // "000..."인 경우
            else -> result
        }
    }
}

File: 프로그래머스/1/131128. 숫자 짝꿍/숫자 짝꿍.js
function solution(X, Y) {
    const xCount = Array(10).fill(0);
    const yCount = Array(10).fill(0);

    // 각 숫자(0~9)의 개수 세기
    for (let ch of X) {
        xCount[parseInt(ch)]++;
    }
    for (let ch of Y) {
        yCount[parseInt(ch)]++;
    }

    let answer = "";

    // 9부터 0까지 반복하며 공통된 개수만큼 추가
    for (let i = 9; i >= 0; i--) {
        const minCount = Math.min(xCount[i], yCount[i]);
        answer += String(i).repeat(minCount);
    }

    if (answer === "") return "-1";
    if (answer[0] === "0") return "0";  // 000...0 인 경우 처리
    return answer;
}
